### DISCUSSION TITLE
xTextBox - writing to a specific character position

### DISCUSSION THREAD

**Discussion by kcobley** [ImpulseAdventure/GUIslice#392] [Q&A]
Hi, first thanks for the excellent library which has already saved me many hours of work!

I'm using a textBox and would like to write a character array to a specific row and column of the text. Is this possible?

I'm trying to produce a field of text that has labels and measured quantities, updating only the measured quantities one at a time in order not to have the display flicker every time its updated. Rewriting the whole field produces very obvious flicker. 

Perhaps there is a better approach than using a textbox for this. 

Many thanks for any advice

Kevin.

**Comment by Pconti31** [comment]
@kcobley If you can post a sample project (just a minimum needed) that reproduces your flicker we might be able to help you out.
IMHO xTextbox won't help you.   Calvin has been working on improvements for when fields are being updated in callbacks, 
[Area of refresh is much larger than the element ](https://github.com/ImpulseAdventure/GUIslice/issues/328)

Paul--

**Reply by kcobley** [reply]
well the code is 500 lines but the part that updates the text was originally like this which cause a lot of flickery text updating - it gets repeated in the main loop:

```
    m_sTextbox1.nBufPosX = 0; 
    m_sTextbox1.nBufPosY = 0;
    sprintf(outString, "Text %+2.1fC Hext %2.0f%%\n", iaqSensor.temperature, iaqSensor.humidity);
    gslc_ElemXTextboxAdd(&m_gui, m_pElemTextbox1, (char*)outString);
    sprintf(outString, "Tsrc %+2.1fC Hsrc %2.0f%%\n", srcTemp.temperature, srcHumidity.relative_humidity);
    gslc_ElemXTextboxAdd(&m_gui, m_pElemTextbox1, (char*)outString);
    sprintf(outString, "Tsup %+2.1fC Hsup %2.0f%%\n", Si7021.readTemperature(), Si7021.readHumidity());
    gslc_ElemXTextboxAdd(&m_gui, m_pElemTextbox1, (char*)outString);
    sprintf(outString, "Pext %4.0fmB IAQ %3.0f  ", iaqSensor.pressure / 100.0, iaqSensor.iaq);
```
And I'm not really sure the first two lines do what I thought they do, as when I tried updating only the numbers, it didn't work at all, with all new output at the end of the last line. I've got a feeling nothing really happens until a newline is sent.  This is partly why I suspect my gui concept is flawed and there is a better way to do it.

The 'doesnt work at all' bit where I tried to overwrite only the numbers:

      ```
  m_sTextbox1.nBufPosX = 17;
        m_sTextbox1.nBufPosY = 2;
        sprintf(outString, "%2.0f",  Si7021.readHumidity());
        gslc_ElemXTextboxAdd(&m_gui, m_pElemTextbox1, (char*)outString);
```
thanks,

Kevin.

**Comment by Pconti31** [comment]
@kcobley Well, without a sample I can't help you.  I have carpal tunnel so I'm not creating one.   You will need to wait for Calvin's input.

If you decide to modify your project as a sample I'd be happy to look at it.   Remember it needs to be stand alone. ie, does not require any outside sensors, just some dummy data.   For example, call readHumidity() and you get random data in the proper range.   

Is it a requirement that the data scroll?
If not, don't use textbox but instead use a page with individual fields and you just update the number fields.
Add a Next and Previous button to show history by jumping to another page that uses either XListbox or XTextbox..
Paul--

**Reply by kcobley** [reply]
sorry - thought what I posted was enough (the excerpt in question that writes to the text box). Sorry to hear you have sore wrists.

I think you have sort of answered my question in that I dont really need scrolling text. So use simple text items instead.

Its quite hard work creating all the tiny snips of text in the GUIbuilder, getting them to align with correct spacing etc., and then after changing the grid spacing to make the lines tight I got this:

![image](https://user-images.githubusercontent.com/15958851/121821500-644ca180-cc91-11eb-9ea1-e6a7487a5b0c.png)

and now my project file is corrupt even after restart :(

such is life! But At least I know what I'm doing now, even though I'm going to have to start the GUI from scratch (and get all the names right etc. so it works with the rest of my code again...)

I find that one always needs to do a project twice in order for it to end up elegant :-)

Thanks for nudging me in the right direction.

**Reply by kcobley** [reply]
Just a note to say that I recovered my project file - the GUISlice Builder app crashed because the grid vertical value got set to zero somehow, and maybe there is no catch for that in the app code. But at least I have my project back again.

**Comment by Pconti31** [comment]
@kcobley I'm happy to help but you might help others if you post a zip file of the crash file and the builder.log files that are in the same folder so I can prevent the crash in the future.

Paul--

**Reply by Pconti31** [reply]
Also, if you read over the User Guide you will find an array of tools for doing alignments.
Paul--

**Comment by kcobley** [comment]
My pleasure - where's the best place to upload the .zip? (I'm not very familiar with Github and normal practice for this stuff)

Looking at the log - it looks like divide by zero error cause by setting the grid vertical to 0 which makes sense. 

I seem to be in good shape now -  using simple text elements. I reduced the effort by having some overlapping parts. The required display is like a table of values. The static parts of the text is now done as four lines of text, and then there is a text field for each of the values to allow them to be written into one at a time.

![image](https://user-images.githubusercontent.com/15958851/121917126-b2fb4980-cd2c-11eb-9517-65e2dee2d086.png)

However I'm having trouble working out how to update the text string in the text element. It doesnt seem to be covered in the documentation. There is discussion of how to make sure the element has a string in RAM, but I cant update that string and get it to show on the screen.

1) How do I reference an element - there is an Enum for it and at object creation there is a reference in prj_GLSC.h but its only a local variable and is overwritten for the next object's creation (This is the auto generated code from GUISliceBuilder, btw.). Is there a way to lookup the ref from the enum?

2) How do I update the text element once I have a ref to it - do I use gslc_ElemSetTxtStr() and does that mark the element as 'touched' for re-drawing?

**Reply by Pconti31** [reply]
@kcobley Thanks for the info.  I found and checked in the fix for zero grids.
If you read the Builder's faq topic `Can you change a Text String at runtime?` that should give you what you need.
As for getting REFERENCES you have to do that by providing a name in the Text element property view for the property **ElementRef**
Example:
```
Key         Text$1
ENUM   	    E_ELEM_HUMIDITY
X           150
Y           100
Width       25
Height      10
ElementRef  m_pHumidity
```
Paul--

**Reply by kcobley** [reply]
Perfect ! I have it working now after a few hiccups. In GUISliceBuilder App, if text is already entered in the text field, and then you change the external storage and ref, the builder doesn't generate the code because it still thinks its static text. Once I deleted the text in the parameters, and set the storage, it correctly switched to dynamic mode showing ??? in the preview.

Text now updates nicely with low flicker (because its not trying to scroll a large area for each line of code) and it looks nice.

For illustations sake:
![image](https://user-images.githubusercontent.com/15958851/122046417-c95ddf80-cdd6-11eb-935a-32e1d7abadcb.png)

Means that the builder autogenerates the following reference in the .ino
```
// Save some element references for direct access
//<Save_References !Start!>
gslc_tsElemRef* Auto              = NULL;
gslc_tsElemRef* Graph             = NULL;
gslc_tsElemRef* High              = NULL;
gslc_tsElemRef* Low               = NULL;
gslc_tsElemRef* Mid               = NULL;
gslc_tsElemRef* OFF               = NULL;
gslc_tsElemRef* Settings          = NULL;
gslc_tsElemRef* m_pElemHextVal    = NULL;
gslc_tsElemRef* m_pElemHsrcVal    = NULL;
gslc_tsElemRef* m_pElemHsupVal    = NULL;
gslc_tsElemRef* m_pElemIAQVal     = NULL; //<--- new element ref
gslc_tsElemRef* m_pElemInTxt2_4   = NULL;
gslc_tsElemRef* m_pElemInTxt3_5   = NULL;
gslc_tsElemRef* m_pElemPextVal    = NULL;
gslc_tsElemRef* m_pElemTextVal    = NULL;
gslc_tsElemRef* m_pElemToggle1_2  = NULL;
gslc_tsElemRef* m_pElemTsrcVal    = NULL;
gslc_tsElemRef* m_pElemTsupVal    = NULL;
gslc_tsElemRef* m_pElemVal1_2_4   = NULL;
gslc_tsElemRef* m_pElemVal1_3     = NULL;
gslc_tsElemRef* m_pElemKeyPadNum  = NULL;
gslc_tsElemRef* m_pElemKeyPadAlpha= NULL;
//<Save_References !End!>
```
Which I used like this in the main loop (note that because its an MKR1000 which is SAMD21, I have full sprintf function):
```
   sprintf(outString, "%3.0f", iaqSensor.iaq);
   gslc_ElemSetTxtStr  ( &m_gui, m_pElemIAQVal ,outString );
```
So I now have my sensors values being reported nicely thanks to your kind help. Next parts of the project are defining the control algortihm, and plotting graphs in the centre portion of the display.  The graphs will be alot of work , as they need dual left right axes with ticks and text labels and whatnot, so will be fresh code as the graph in GUISlice is fairly simple.

However, the bare bones are there now. Thank-you.
